utils/heatmaper.py

```diff
@@ -0,0 +1,171 @@
+import os
+import sys
+import torch
+import cv2
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+sys.path.append(os.path.dirname(os.path.abspath(__file__)) + '/../')
+from data.utils import transform # noqa
+from model.utils import tensor2numpy # noqa
+
+
+disease_classes = [
+    'Cardiomegaly',
+    'Edema',
+    'Consolidation',
+    'Atelectasis',
+    'Pleural Effusion'
+]
+
+
+def fig2data(fig):
+    fig.canvas.draw()
+    # Get the RGB buffer from the figure
+    w, h = fig.canvas.get_width_height()
+    buf = np.fromstring(fig.canvas.tostring_rgb(), dtype=np.uint8)
+    buf.shape = (h, w, 3)
+    buf = buf[:, :, (2, 1, 0)]
+    return np.ascontiguousarray(buf)
+
+
+class Heatmaper(object):
+    def __init__(self, alpha, prefix, cfg, model, device):
+        super(Heatmaper, self).__init__()
+        # init cfg
+        self.cfg = cfg
+        # init device
+        self.device = device
+        # init model
+        self.model = model
+        # init transparency
+        self.alpha = alpha
+        # init filename prefix
+        self.prefix = prefix
+
+    def image_reader(self, image_file):
+        """
+        Args:
+            image_file: str to an image file path
+        Returns:
+            image: (1, C, H, W) tensor of image
+            image_color: (H, W, C) numpy array of BGR image
+        """
+        image_gray = cv2.imread(image_file, 0)
+        assert image_gray is not None, "invalid image read in: {}"\
+            .format(image_file)
+        image_color = cv2.cvtColor(image_gray, cv2.COLOR_GRAY2BGR)
+        image = transform(image_gray, self.cfg)
+        image = torch.from_numpy(image)
+        image = image.unsqueeze(0)
+        return image, image_color
+
+    def set_overlay(self, figure, row_, i, subtitle):
+        ax_overlay = figure.add_subplot(row_, 3, 2 + i)
+        ax_overlay.set_title(subtitle, fontsize=10, color='r')
+        ax_overlay.set_yticklabels([])
+        ax_overlay.set_xticklabels([])
+        ax_overlay.tick_params(axis='both', which='both', length=0)
+        return ax_overlay
+
+    def set_rawimage(self, figure, row_):
+        ax_raw_image = figure.add_subplot(row_, 3, 1)
+        subtitle = 'original image'
+        ax_raw_image.set_title(subtitle, fontsize=10, color='r')
+        ax_raw_image.set_yticklabels([])
+        ax_raw_image.set_xticklabels([])
+        ax_raw_image.tick_params(axis='both', which='both', length=0)
+        return ax_raw_image
+
+    def get_raw_image(self, image, ax):
+        """
+        Args:
+            image: (H, W, C) numpy array of BGR image
+            ax: AxesSubplot of matplotlib.axes
+        Returns:
+            image: AxesImage of matplotlib.image
+        """
+        image = ax.imshow(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))
```

Comment:
assert image is BGR?  you have transfer image gray to BGR  at the line of 57, and BGR to RGB here.

Reply:
i don't get what you mean. Please use declarative sentence if you do not have questions here.

Reply:
> assert image is BGR? you have transfer image gray to BGR at the line of 57, and BGR to RGB here.

I can answer your questions now. Two times of transform image  because of this: the first transform at the line of 57 where convert `GRAY` to `BGR`  is aimed to stroe the image as opencv image format, and the second one is aimed to use the image as matplotlib format.  It's better to understand the code and debug, i think.

Reply:
If you're going to use this variable, of course, you should convert it to the OpenCV image format, but you haven't used this variable in the middle, so I don't think you need to convert it twice in a row. Consolidation is recommended to increase computing speed.